tools/base/checker.py

```diff
@@ -0,0 +1,157 @@
+import argparse
+import logging
+import os
+import subprocess
+import sys
+from functools import cached_property
+
+LOG_LEVELS = (("info", logging.INFO), ("debug", logging.DEBUG), ("warn", logging.WARN),
+              ("error", logging.ERROR))
+
+
+class Checker(object):
+    """Runs check methods prefixed with `check_` and named in `self.checks`
+
+    check methods should return the count of errors and log warnings and errors
+    """
+    checks = ()
+
+    def __init__(self, *args):
+        self._args = args
+        self._checks_run = {}
+        self.errors = {}
+        self.warnings = {}
+
+    @cached_property
+    def args(self):
+        return self.parser.parse_args(self._args)
+
+    @cached_property
+    def log_level(self):
+        return dict(LOG_LEVELS)[self.args.log_level]
+
+    @property
+    def diff(self):
+        return self.args.diff
+
+    @property
+    def fix(self):
+        return self.args.fix
+
+    @cached_property
+    def log(self):
+        logger = logging.getLogger(__name__)
+        logger.setLevel(self.log_level)
+        stderr_handler = logging.StreamHandler(sys.stderr)
+        stderr_handler.setLevel(logging.WARN)
+        stdout_handler = logging.StreamHandler(sys.stdout)
+        stdout_handler.setLevel(logging.DEBUG)
```

Comment:
calling main on external libs without a subproc *can* mess with this - need to think this through a bit